File: src/test/matrix/py/progress.py
#!/usr/bin/env python3
"""MCP progress matrix cell server (feature=progress, language=py, transport=stdio).

Feature under test: PROGRESS NOTIFICATIONS DURING A TOOL CALL.

When a tools/call arrives carrying params._meta.progressToken, the tool handler
emits several notifications/progress messages (progress increasing toward total)
before it returns its final result.

How this is wired:
  FastMCP populates ctx.request_context.meta.progressToken from the inbound
  request's params._meta.progressToken. ctx.report_progress(progress, total,
  message) then turns into a server->client notifications/progress message:
      {"jsonrpc":"2.0","method":"notifications/progress",
       "params":{"progressToken":<token>,"progress":<n>,"total":<t>,"message":...}}
  If NO progressToken was supplied, report_progress is a no-op (per spec: the
  server must not send progress for a token the client didn't ask for).

Exposes:
  - 1 progress tool:  long_task  (emits 4 progress notifications, then returns)
  - 1 plain tool:     echo       (no progress; keeps a trivial tool available)

Expected behavior against mcpgaze:
  initialize + tools/list succeed (snapshot lists 2 tools, conform --all PASSES,
  health --once is UP). When record/wrap drive a tools/call with
  params._meta.progressToken set, the captured session log / cassette MUST
  contain the notifications/progress messages, well-formed, with the matching
  progressToken and monotonically increasing progress values. If mcpgaze drops
  or garbles those notifications, that is a real mcpgaze bug.

Built on the real Python MCP SDK (FastMCP).
"""

import anyio
from mcp.server.fastmcp import Context, FastMCP

mcp = FastMCP("mcpgaze-progress-py")

# Number of progress steps emitted by long_task before it returns.
STEPS = 4


# --- Progress tool: emits several progress notifications before returning ---
@mcp.tool(
    name="long_task",
    title="Run a long task with progress",
    description=(
        "Run a multi-step task that emits notifications/progress updates "
        "(progress increasing toward a total) while it works, then returns a "
        "completion string. Progress is only emitted when the caller supplies "
        "params._meta.progressToken."
    ),
)
async def long_task(label: str, ctx: Context) -> str:
    """Emit STEPS progress notifications (1..STEPS of STEPS), then finish.

    report_progress is a no-op unless the inbound request carried a
    progressToken, so this tool is safe to call without one as well.
    """
    for step in range(1, STEPS + 1):
        await ctx.report_progress(
            progress=step,
            total=STEPS,
            message=f"{label}: step {step}/{STEPS}",
        )
        # Tiny yield so the notifications are flushed as distinct wire frames
        # rather than coalesced; keeps the run fast for the smoke test.
        await anyio.sleep(0.01)
    return f"{label}: completed {STEPS}/{STEPS} steps"


# --- Plain tool (no progress) ---------------------------------------------
@mcp.tool()
def echo(message: str) -> str:
    """Echo back the provided message (no progress involved)."""
    return f"echo: {message}"


if __name__ == "__main__":
    mcp.run(transport="stdio")

File: src/test/matrix/ts/progress.mjs
#!/usr/bin/env node
// progress-ts matrix cell: MCP server (stdio) whose tool emits a sequence of
// notifications/progress messages (progress increasing toward total) BEFORE it
// returns the tool result. Uses the real @modelcontextprotocol/sdk.
//
// initialize + tools/list are always normal. The progress tool only emits
// progress notifications when the CLIENT supplies params._meta.progressToken
// on the tools/call request (per the MCP spec, a server MUST NOT send progress
// for a token it was never given). The SDK surfaces that token to the tool
// handler via the `extra` argument:
//   - extra._meta.progressToken  -> the token to echo back
//   - extra.sendNotification(..) -> emits a notification correlated to this call
// We send the notification at the protocol method level so it goes on the wire
// verbatim as { method:"notifications/progress", params:{ progressToken,
// progress, total, message } }.
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";

const server = new McpServer(
  { name: "progress-ts", version: "1.0.0" },
  { capabilities: { tools: {} } }
);

const TOTAL = 4;

// The progress tool: emits TOTAL progress notifications (1..TOTAL of TOTAL)
// before returning a final result. If no progressToken was supplied by the
// client, it skips the notifications (spec-compliant) and just returns.
server.registerTool(
  "long_task",
  {
    title: "Long-running task with progress",
    description:
      "Runs a simulated multi-step task. When the caller supplies a " +
      "progressToken in params._meta, emits notifications/progress messages " +
      "(progress increasing toward total) before returning the final result.",
    inputSchema: {
      steps: z
        .number()
        .int()
        .positive()
        .optional()
        .describe("Number of progress steps to emit (default 4)"),
    },
  },
  async ({ steps }, extra) => {
    const total = Number.isInteger(steps) && steps > 0 ? steps : TOTAL;
    const progressToken = extra?._meta?.progressToken;

    if (progressToken !== undefined && progressToken !== null) {
      for (let i = 1; i <= total; i++) {
        // Emit a well-formed progress notification correlated to this call.
        await extra.sendNotification({
          method: "notifications/progress",
          params: {
            progressToken,
            progress: i,
            total,
            message: `step ${i} of ${total}`,
          },
        });
      }
    }

    return {
      content: [
        {
          type: "text",
          text:
            progressToken !== undefined && progressToken !== null
              ? `completed ${total} steps (emitted ${total} progress notifications)`
              : `completed ${total} steps (no progressToken supplied; no notifications emitted)`,
        },
      ],
    };
  }
);

// A plain tool so tools/list is meaningful even without progress.
server.registerTool(
  "echo",
  {
    title: "Echo",
    description: "Echo back the provided message (no progress involved)",
    inputSchema: { message: z.string().describe("Message to echo back") },
  },
  async ({ message }) => ({
    content: [{ type: "text", text: `echo: ${message}` }],
  })
);

const transport = new StdioServerTransport();
await server.connect(transport);
